rj_gameplay/stp/role/cost.py

```diff
@@ -0,0 +1,72 @@
+from typing import Optional
+
+import numpy as np
+
+import stp.role as role
+import stp.game_state as game_state
+
+
+def flat_switch_cost(
+    robot: game_state.Robot, prev_robot: Optional[game_state.Robot], switch_cost: float
+) -> float:
+    if prev_robot:
+        return switch_cost * (robot.id != prev_robot.id)
+    else:
+        return 0
+
+
+def distance_to_pt(
+    pt: np.ndarray, saturate_dist: float, switch_cost: float
+) -> role.CostFn:
+    """Creates a cost function that returns the distance from the robot's position to a
+    point, saturating at the passed in distance.
+    :param pt: Point to evaluate robot's position to.
+    :param saturate_dist: Distance at which the cost function saturates.
+    :param switch_cost: Flat cost added for switching to a different robot.
+    :return: Cost function
+    """
+
+    def cost_fn(robot: game_state.Robot, prev_robot: Optional[game_state.Robot]):
+        dist: float = np.linalg.norm(robot.pose[:2] - pt)
+        cost: float = dist / saturate_dist
+
+        # Add the cost for switching robots.
+        cost += flat_switch_cost(robot, prev_robot, switch_cost)
+
+        return max(min(cost, 1.0), 0.0)
```

Comment:
Are we intending to limit all of our costs to the (0, 1) range? If so, this should happen in a "soft" way (i.e. by tweaking constants so it happens like that in normal operating conditions) rather than by hard-limiting. This just loses signal and makes it more confusing to combine things later.

Reply:
I think it would be nice to limit all of our costs to [0, 1], just so it's easier to compare costs or something. I'll need to think about it more about what the best way of doing this is.

Reply:
I personally really hate limiting costs - it tends to just mask over the weird edge cases. I would prefer to express all costs in (approximate) seconds or something, that tends to make it easier to reason about.

Reply:
What do you mean by (approximate) seconds?

Reply:
I mean like we should scale them to a physically intuitive unit. Seconds was my first thought, because it's pretty easy to define and reason about.
My preferred definition of cost would be, roughly, "time until the robot is able to start actively participating in this role". For example, for a capture skill, this would equate to "time to capture the ball". In this context, I used "approximate" because obviously the cost function won't know _exactly_ how long it is.

Reply:
Bump on this?

Reply:
Done.